app/lib/crypto/address/p2pkh.ts

```diff
@@ -22,8 +22,30 @@ export const encode = (pubKey: PublicKey, chain: ChainType): string => {
   if (!(chain in Prefix)) {
     throw new Error(Errors.UNSUPPORTED_CHAIN_TYPE)
   }
-  const b32 = Bech32.toWords(Buffer.concat([Buffer.from([0]), sha256(pubKey.bytes).slice(20)]))
+  const b32 = Bech32.toWords(Buffer.concat([Buffer.from([0]), sha256(pubKey.bytes).slice(0, 20)]))
   const hrp = Prefix[chain]
 
   return Bech32.encode(hrp, b32)
-}
\ No newline at end of file
+}
+
+/**
+ * Decode the chain type and the hash of the first 20 bytes of the public key
+ * @param {string} address
+ * @returns {[ChainType , Buffer]}
+ */
+export const decode = (address: string): [ChainType, Buffer] => {
+  const {prefix, words} = Bech32.decode(address)
+  const keyHash = Buffer.from(Bech32.fromWords(words))
+
+  let chainCode: ChainType
+
+  if (prefix === "twit") {
+    chainCode = ChainType.test
+  } else if (prefix === "wit") {
+    chainCode = ChainType.main
+  } else {
+    throw new Error(Errors.UNSUPPORTED_CHAIN_TYPE)
+  }
```

Comment:
Making the most of the magic properties of enums, behold this:
```ts
// Unless `ChainTypes[...]` is used somewhere else, this enum can be made constant.
const enum ChainTypes {
  test,
  main
}

// Note that here I'm making the mapping between ChainTypes and Prefixes explicitly, jut in case
// we change the members order in any of the enums at some point, which would remain unnoticed
// by the compiler and would break all logic.
enum Prefixes {
  twit = ChainTypes.test,
  wit = ChainTypes.main
}

// As Prefixes is not a constant enum, it can be easily queried for member pertenence.
if (!(prefix in Prefixes)) {
  throw new Error(Errors.UNSUPPORTED_CHAIN_TYPE)
}

// This is perfectly safe and the compiler recognizes the type as `ChainType`.
const chainType = Prefixes[prefix]
```